test/infrastructure/inmemory/internal/controllers/inmemorymachine_controller.go

```diff
@@ -181,28 +195,125 @@ func (r *InMemoryMachineReconciler) reconcileNormal(ctx context.Context, cluster
 		return ctrl.Result{}, nil
 	}
 
-	// Create VM
+	// Call the inner reconciliation methods.
+	phases := []func(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, inMemoryMachine *infrav1.InMemoryMachine) (ctrl.Result, error){
+		r.reconcileNormalCloudMachine,
+		r.reconcileNormalNode,
+		r.reconcileNormalETCD,
+		r.reconcileNormalAPIServer,
+		r.reconcileNormalScheduler,
+		r.reconcileNormalControllerManager,
+		r.reconcileNormalKubeadmObjects,
+	}
+
+	res := ctrl.Result{}
+	errs := []error{}
+	for _, phase := range phases {
+		phaseResult, err := phase(ctx, cluster, machine, inMemoryMachine)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if len(errs) > 0 {
+			continue
+		}
+		// TODO: consider if we have to use max(RequeueAfter) instead of min(RequeueAfter) to reduce the pressure on
+		//  the reconcile queue for InMemoryMachines given that we are requeuing just to wait for some period to expire;
+		//  the downside of it is that InMemoryMachines status will change by "big steps" vs incrementally.
+		res = util.LowestNonZeroResult(res, phaseResult)
+	}
+	return res, kerrors.NewAggregate(errs)
+}
+
+func (r *InMemoryMachineReconciler) reconcileNormalCloudMachine(ctx context.Context, cluster *clusterv1.Cluster, _ *clusterv1.Machine, inMemoryMachine *infrav1.InMemoryMachine) (ctrl.Result, error) {
+	// Compute the resource group unique name.
+	// NOTE: We are using reconcilerGroup also as a name for the listener for sake of simplicity.
+	resourceGroup := klog.KObj(cluster).String()
+	cloudClient := r.CloudManager.GetResourceGroup(resourceGroup).GetClient()
+
+	// Create VM; a Cloud VM can be created as soon as the Infra Machine is created
 	// NOTE: for sake of simplicity we keep cloud resources as global resources (namespace empty).
-	// TODO: we should convert this into a full reconcile (if it exist, update)
 	cloudMachine := &cloudv1.CloudMachine{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: inMemoryMachine.Name,
 		},
 	}
-	if err := cloudClient.Create(ctx, cloudMachine); err != nil && !apierrors.IsAlreadyExists(err) {
-		return ctrl.Result{}, errors.Wrapf(err, "failed to create CloudMachine")
+	if err := cloudClient.Get(ctx, client.ObjectKeyFromObject(cloudMachine), cloudMachine); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		if err := cloudClient.Create(ctx, cloudMachine); err != nil && !apierrors.IsAlreadyExists(err) {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create CloudMachine")
+		}
+	}
+
+	// Wait for the VM to be provisioned; provisioned happens a configurable time after the cloud machine creation.
+	provisioningDuration := time.Duration(0)
+	if inMemoryMachine.Spec.Behaviour != nil && inMemoryMachine.Spec.Behaviour.VM != nil {
+		x := inMemoryMachine.Spec.Behaviour.VM.Provisioning
+
+		provisioningDuration = x.StartupDuration.Duration
+		jitter, _ := strconv.ParseFloat(x.StartupJitter, 64)
```

Comment:
I would prefer returning an error in this case. If someone writes an invalid float it will be hard to detect otherwise

Same below in other cases